crypto/thread/arch/thread_win.c

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include "common.h"
+
+#if defined(OPENSSL_THREADS_WINNT)
+# include <process.h>
+# include <internal/thread_arch.h>
+
+static DWORD __stdcall thread_start_thunk(LPVOID vthread)
+{
+    void *cb_data;
+    void (*cb)(void*);
+    CRYPTO_THREAD *thread;
+    CRYPTO_THREAD_RETVAL ret;
+    cb = NULL;
```

Comment:
You're right, they are rather similar.

The reason why I chose to leave them separate:
 - I like the idea of matching the signature required by the posix/crt for these kinds of API calls for clarity, i.e., `void* foo(void *);` and `DWORD __stdcall foo(LPVOID)`.
 - There are bits and pieces that are specific to pthread/crt.

So we could either use on this (quite ugly and unreadable, imo) preprocessor approach:

```c
#if defined(OPENSSL_THREADS_POSIX)
static void *thread_start_thunk(void *vthread)
#elif defined(OPENSSL_THREADS_WINNT)
static DWORD __stdcall thread_start_thunk(LPVOID vthread)
#endif
{
    // ...
}
```
or use something like
```c
#if defined(OPENSSL_THREADS_POSIX)
typedef void * THUNK_RETVAL;
#define THUNK_CC
#elif defined(OPENSSL_THREADS_WINNT)
typedef DWORD THUNK_RETVAL;
#define THUNK_CC __stdcall
#endif

static THUNK_RETVAL THUNK_CC thread_start_thunk(LPVOID vthread)
```
which obscures the signature (on first glance) and we still need to use preprocessor (or static inlines for):
```c
{
    // ...
#if defined(OPENSSL_THREADS_POSIX)
    pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
    pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
#elif defined(OPENSSL_THREADS_WINNT)
    thread->thread_id = GetCurrentThreadId();
#endif
    // ...
#if defined(OPENSSL_THREADS_POSIX)
    return NULL;
#elif defined(OPENSSL_THREADS_WINNT)
    return 0;
#endif
}
```

If we'd like to dedup the code, rather than merging the two functions, it might make more sense to split `thread_start_thunk` into `thread_start_thunk` and `thread_start_thunk_common` :
```c
// common.h
static ossl_inline void thread_start_thunk_common(CRYPTO_THREAD *thread)
{
    void *cb_data;
    void (*cb)(void*);
    CRYPTO_THREAD_RETVAL ret;

    cb = NULL;
    cb_data = NULL;
    ret = thread->routine(thread->data, &cb, &cb_data);

    CRYPTO_MUTEX_lock(thread->statelock);
    CRYPTO_THREAD_SET_STATE(thread, CRYPTO_THREAD_FINISHED);
    thread->retval = ret;
    CRYPTO_CONDVAR_broadcast(thread->condvar);
    CRYPTO_MUTEX_unlock(thread->statelock);

    if (cb != NULL)
        cb(cb_data);
}

// thread_win.c
static DWORD __stdcall thread_start_thunk(LPVOID vthread)
{
    CRYPTO_THREAD *thread;

    thread = (CRYPTO_THREAD *) vthread;
    thread->thread_id = GetCurrentThreadId();

    thread_start_thunk_common(thread);
    return 0;
}

// thread_posix.c
static void *thread_start_thunk(void *vthread)
{
    CRYPTO_THREAD *thread;

    thread = (CRYPTO_THREAD *) vthread;

    pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, NULL);
    pthread_setcanceltype(PTHREAD_CANCEL_ASYNCHRONOUS, NULL);

    thread_start_thunk_common(thread);
    return NULL;
}
```

WDYT?

Reply:
Sorry for the misunderstanding! Merged init and decl.